RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducer.cc

```diff
@@ -8,55 +8,87 @@
 #include "RecoParticleFlow/PFRecHitProducer/interface/PFRecHitTopologyRecord.h"
 
 namespace ALPAKA_ACCELERATOR_NAMESPACE {
-  class PFClusterSoAProducer : public stream::EDProducer<> {
+  class PFClusterSoAProducer : public stream::SynchronizingEDProducer<> {
   public:
     PFClusterSoAProducer(edm::ParameterSet const& config)
         : pfClusParamsToken(esConsumes(config.getParameter<edm::ESInputTag>("pfClusterParams"))),
           topologyToken_(esConsumes(config.getParameter<edm::ESInputTag>("topology"))),
           inputPFRecHitSoA_Token_{consumes(config.getParameter<edm::InputTag>("pfRecHits"))},
           outputPFClusterSoA_Token_{produces()},
           outputPFRHFractionSoA_Token_{produces()},
+          num_rhf_{cms::alpakatools::make_host_buffer<uint32_t>()},
           synchronise_(config.getParameter<bool>("synchronise")),
           pfRecHitFractionAllocation_(config.getParameter<int>("pfRecHitFractionAllocation")) {}
 
-    void produce(device::Event& event, device::EventSetup const& setup) override {
+    void acquire(device::Event const& event, device::EventSetup const& setup) override {
       const reco::PFClusterParamsDeviceCollection& params = setup.getData(pfClusParamsToken);
       const reco::PFRecHitHCALTopologyDeviceCollection& topology = setup.getData(topologyToken_);
       const reco::PFRecHitHostCollection& pfRecHits = event.get(inputPFRecHitSoA_Token_);
       int nRH = 0;
       if (pfRecHits->metadata().size() != 0)
         nRH = pfRecHits->size();
 
-      reco::PFClusteringVarsDeviceCollection pfClusteringVars{nRH, event.queue()};
-      reco::PFClusteringEdgeVarsDeviceCollection pfClusteringEdgeVars{(nRH * 8), event.queue()};
-      reco::PFClusterDeviceCollection pfClusters{nRH, event.queue()};
-      reco::PFRecHitFractionDeviceCollection pfrhFractions{nRH * pfRecHitFractionAllocation_, event.queue()};
+      pfClusteringVars = std::make_unique<reco::PFClusteringVarsDeviceCollection>(nRH, event.queue());
+      pfClusteringEdgeVars = std::make_unique<reco::PFClusteringEdgeVarsDeviceCollection>(nRH * 8, event.queue());
+      pfClusters = std::make_unique<reco::PFClusterDeviceCollection>(nRH, event.queue());
+
+      // device buffer to store rhf size
+      auto size_d = cms::alpakatools::make_device_buffer<uint32_t>(event.queue());
+      alpaka::memset(event.queue(), size_d, 0x00);
+
+      if (nRH != 0) {
+        PFClusterProducerKernel kernel(event.queue(), pfRecHits);
+        kernel.step1(event.queue(),
+                     params,
+                     topology,
+                     *pfClusteringVars,
+                     *pfClusteringEdgeVars,
+                     pfRecHits,
+                     *pfClusters,
+                     size_d.data());
+      }
+
+      auto device_num_rhf = cms::alpakatools::make_device_view<uint32_t>(alpaka::getDev(event.queue()), *size_d.data());
+
+      alpaka::memcpy(event.queue(), num_rhf_, device_num_rhf);
```

Comment:
@jsamudio if you fell like trying something new, I would be interested to hear your feedback about a slightly different approach.

If you declare `num_rhf_` as a host buffer in pinned memory (see above), the kernel can write directly to it, avoiding the need for the extra `memcpy` operation:
```suggestion
      *num_rhf_ = 0;

      if (nRH != 0) {
        PFClusterProducerKernel kernel(event.queue(), pfRecHits);
        kernel.step1(event.queue(),
                     params,
                     topology,
                     *pfClusteringVars,
                     *pfClusteringEdgeVars,
                     pfRecHits,
                     *pfClusters,
                     num_rhf_.data());
      }
```

Reply:
This worked well, and the change was made.